### PR TITLE
Transfer-Encoding takes precedence over Content-Length.

### PR DESCRIPTION
**Pull request by rzarzynski**
This patch changes the way how misbehaving clients, that send `Content-Length` together with `Transfer-Encoding`, are handled. Currently we don't do dechunking but apply the processing for messages with valid `Content-Length`. Unfortunately, this brings significant inconsistency with S3 as its affects AWSv4.
The fix brings the same mechanism we have in the `master` branch of `ceph/civetweb` repo.

Fixes: http://tracker.ceph.com/issues/21015
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

### CODE REVIEW

**Review by mattbenjamin — COMMENTED** [review]
lgtm, marcus?

**Comment by alfredodeza** [comment]
Since this is a fork, we usually require upstream patches first, and then we pull them in (vs. accepting changes directly to the fork)

**Comment by ktdreyer** [comment]
Would you please link to the PR to Civetweb upstream for this?

**Comment by rzarzynski** [comment]
Guys, I see your concerns. At the moment there is no related PR in the upstream. Though, it's also worth to note the same algorithm is present in the `master` of `ceph/civetweb` for over 3 years. The priority for `Transfer-Encoding` has been introduced on 15 Aug 2014 in https://github.com/ceph/civetweb/commit/2cd3759f97d3392fba64d1f999b1d2433a96bff8.

File: src/civetweb.c [hunk 1]
```diff
@@ -11955,7 +11955,12 @@ getreq(struct mg_connection *conn, char *ebuf, size_t ebuf_len, int *err)
 		return 0;
 	} else {
 		/* Message is a valid request or response */
-		if ((cl = get_header(&conn->request_info, "Content-Length")) != NULL) {
+		if ((cl = get_header(&conn->request_info, "Transfer-Encoding"))
+		               != NULL
+		           && !mg_strcasecmp(cl, "chunked")) {
+			conn->is_chunked = 1;
+			conn->content_len = 0;
```
Comment (yehudasa) [inline src/civetweb.c:11962]: @rzarzynski the original block copied doesn't set `content_length`, is there a reason why you do that?

Comment (rzarzynski) [reply src/civetweb.c:11962]: This is because  similar change (c39d732637a813ab5d4386a6167444dd6a235e82) has been introduced to `master`.

Comment (yehudasa) [reply src/civetweb.c:11962]: it certainly has. I'm not sure what's the status of `master` though, and why this hasn't made its way to this branch earlier.

**Comment by alexiri** [comment]
Hi @rzarzynski, is there an upstream PR for this change? I see you made the [changes](https://github.com/rzarzynski/civetweb/commit/244aa8a03ed6c5eb959a396955a7aa3e01667a97) to the code already, but I can't find the PR.

**Comment by rzarzynski** [comment]
Hello @alexiri! The part related to `content_len` handling has been merged in https://github.com/civetweb/civetweb/pull/511.

**Comment by theanalyst** [comment]
running tests at http://pulpito.ceph.com/abhi-2017-10-24_13:57:15-rgw-wip-civetweb-fixes-distro-basic-smithi/

**Pull request merged**